### PR TITLE
Fix Flex install script for Flex 4.14+ and Feathers SDK

### PR DESCRIPTION
**Pull request by matyasf**
Flex 4.14 introduced new properties for its install ant script; if these properties are not set the proprietary components (AIR SDK, playerGlobal,..) will not be downloaded and thus the SDK will be unusable for GradleFX. This PR fixes this issue.

Note: This was actually a bugfix by the Flex team, the installed was supposed to ask for these properties just because of a bug it did not do so. For details see https://github.com/apache/flex-sdk/commit/67f1f5d3c522c723e0325b3a4aed68e518cb5e3b  

Note2: There is a new SDK for Starling called Feathers SDK that has similar structure as Flex 4.14, this commit makes it work with GradleFx.

### CODE REVIEW

File: src/main/groovy/org/gradlefx/configuration/sdk/states/flex/InstallFlexSdkState.groovy [hunk 1]
```diff
@@ -97,6 +101,9 @@ class InstallFlexSdkState extends AbstractInstallSdkState {
             if(!showPrompts) {
                 property(name: 'installer', value: true)
             }
+            property(name: 'do.air.install', value: true)
+            property(name: 'do.swfobject.install', value: true)
+            property(name: 'do.fontswf.install', value: true)
```
Comment (matyasf) [inline src/main/groovy/org/gradlefx/configuration/sdk/states/flex/InstallFlexSdkState.groovy]: This is the most important part, the actual fix.

File: src/main/groovy/org/gradlefx/configuration/sdk/states/AbstractCreateSdkInstallLocationState.groovy [hunk 2]
```diff
@@ -32,14 +32,14 @@ abstract class AbstractCreateSdkInstallLocationState implements SdkInitState {
     SdkInstallLocation installLocation
     String sdkFile
     protected String configName
-    Boolean isInstallationRequired
+    Boolean hasSDKDeclaredAsDependency
```
Comment (matyasf) [inline src/main/groovy/org/gradlefx/configuration/sdk/states/AbstractCreateSdkInstallLocationState.groovy]: this is just a small refactoring to make this part of the code more understandable.

File: src/main/groovy/org/gradlefx/configuration/sdk/states/flex/InstallFlexSdkState.groovy [hunk 3]
```diff
@@ -97,6 +101,10 @@ class InstallFlexSdkState extends AbstractInstallSdkState {
             if(!showPrompts) {
                 property(name: 'installer', value: true)
             }
+            property(name: 'do.air.install', value: true)
+            property(name: 'do.flash.install', value: true)
+            property(name: 'do.swfobject.install', value: true)
+            property(name: 'do.fontswf.install', value: true)
```
Comment (matyasf) [inline src/main/groovy/org/gradlefx/configuration/sdk/states/flex/InstallFlexSdkState.groovy]: this is the actual fix, the most important part

**Comment by SlevinBE** [comment]
Thanks for the PR!
I would not download the AIR SDK though, as this will download a default version. This breaks the functionality we have right now where you can specify a specific AIR SDK as a depedency (GradleFx thinks an AIR SDK has already been installed, but it's actually the default one so it won't try to install the one specified as a dependency)

**Comment by matyasatfp** [comment]
hmhm are you sure?

The current behaviour does download the AIR SDK when you define just Flex SDK < 4.14 as a dependency. Look at the install script that I linked in the Flex changeset, no matter what you do if you run that script it will install them.        
And as I know if I only specify Flex 4.14 as a dependency build will fail since it wont download playerGlobal/airglobal.   

But I might be wrong, I will double check

**Comment by SlevinBE** [comment]
Maybe I made a mistake during my test, but I tried to build a project with a 4.14.1 Flex SDK dependency and an AIR 18.0 SDK. What I saw was that the Flex SDK was being installed, but that it skipped the AIR 18.0 SDK since an earlier version (16.0) was already installed by default during the Flex SDK installation.

**Comment by SlevinBE** [comment]
@matyasatfp So I re-did my test to make sure I didn't f*ck up :D

First I only specified Flex SDK 4.14.0 as a dependency, without your fix. What I saw was that only the Flex SDK was being installed. However, there are some AIR files included by default, such as the frameworks/air-config.xml file and 'frameworks/libs/air' folder. But this isn't the complete AIR SDK, as no 'lib/adt.jar' and other required files are installed to create an AIR package.

Then I specified the Flex SDK 4.14.1 with your GradleFx version. Then I saw that the complete AIR SDK was being downloaded, with all the required AIR files such as 'lib/adt.jar'.

PS: can you also contact me via the contact details on my profile (https://github.com/SlevinBE). I've got another question to ask you ;)

**Comment by matyasf** [comment]
Just a note, I am still looking into this. This PR is not perfect, if I figure out what is exactly happening I will add a commit. 

This is the result of my testing on a Flex project (with Gradle 2.1) without my patch:

Just Flex 4.13.0 as dependency: 
    flexSDK group: 'org.apache', name: 'apache-flex-sdk', version: '4.13.0', ext: 'tar.gz'    
Installs Flex correctly with the bundled 16.0 AIR SDK

Flex 4.13.0 + AIR 18:   
    flexSDK group: 'org.apache', name: 'apache-flex-sdk', version: '4.13.0', ext: 'tar.gz'    
    airSDK group: 'com.adobe', name: 'AIRSDK', version: '18.0', ext: Os.isFamily(Os.FAMILY_WINDOWS) ? 'zip' : 'tbz2'    
Does not overlay AIR 18, uses AIR 16

Just Flex 4.14.1:   
    flexSDK group: 'org.apache', name: 'apache-flex-sdk', version: '4.14.1', ext: 'tar.gz'    
Does not work, AIR SDK is not installed. flexconfig.xml looks correct.    

Flex 4.14.1 + AIR 18:      
    flexSDK group: 'org.apache', name: 'apache-flex-sdk', version: '4.14.1', ext: 'tar.gz'      
    airSDK group: 'com.adobe', name: 'AIRSDK', version: '18.0', ext:   Os.isFamily(Os.FAMILY_WINDOWS) ? 'zip' : 'tbz2'     
While AIR is overlayed correctly, AIR installation seems to overwrite flexconfig.xml so it wont have entries to include the Flex SDK. Also font encoding does not work either.

Or am I doing something wrong?   

I will investigate it more, and try to fix the issues.    
My main goal is to get the Flex 4.14.1 + AIR 18 case working for pure AS3 projects (= fix the font encoding jars not added), since this is what I think will make the Feathers SDK work.

File: src/main/groovy/org/gradlefx/configuration/sdk/states/flex/InstallFlexSdkState.groovy [hunk 4]
```diff
@@ -47,6 +47,10 @@ class InstallFlexSdkState extends AbstractInstallSdkState {
             downloadPlayerGlobalSwc()
             updateFrameworkConfigFiles()
         }
+        else {
+            throw new RuntimeException("Unable to install Flex SDK. No install script found. The script needs either" +
+                    "${getInstallerScriptFile().path} or ${getAdditionalDownloadsAntScriptFile().path} in the SDK.")
+        }
```
Comment (jlopez) [inline src/main/groovy/org/gradlefx/configuration/sdk/states/flex/InstallFlexSdkState.groovy]: the problem with this is that older Flex SDKs will fail to install...

**Comment by SlevinBE** [comment]
The issue I encountered previously in this PR (AIR dependency not being overlayed) is probably solved by the sentry file mechanism introduced with https://github.com/GradleFx/GradleFx/pull/218 by @jlopez 

So I'll give this another go when I find some time.

**Comment by matyasf** [comment]
I have tested GradleFx 1.3.3 on a mobile project and a Flash project, both seem to work. Meaning that this PR is not needed.

**Pull request closed without merging**